File: agent-plane/src/alphamintx_agent_plane/backtest/dataset.py
"""Canonical backtest dataset (JSONL klines) parsing, validation, and slicing.

The dataset file is produced by the Go ``backtestctl fetch`` step and consumed
read-only here (docs/specs/backtest-engine.md): one kline per LF-terminated
compact-JSON line, a single (symbol, interval) per file, strictly ascending
grid-aligned ``open_time`` (ms epoch), OHLCV as decimal strings (contract
DecimalStr rules, ADR-0003). Gaps in the grid are LEGAL; duplicates,
out-of-order, misaligned, or mixed rows are validation errors.
``close_time`` is never stored — it is derived as ``open_time + d_ms``.
``dataset_sha256`` is the sha256 hex of the ENTIRE file bytes.
"""

from __future__ import annotations

import hashlib
import json
from pathlib import Path

from pydantic import BaseModel, ConfigDict, Field, ValidationError

from alphamintx_agent_plane.contract.models import DecimalStr, SymbolStr, decimal_to_str

# The only legal dataset intervals (cross-plane contract interval table).
INTERVAL_SECONDS: dict[str, int] = {
    "1m": 60,
    "3m": 180,
    "5m": 300,
    "15m": 900,
    "30m": 1800,
    "1h": 3600,
    "2h": 7200,
    "4h": 14400,
    "6h": 21600,
    "8h": 28800,
    "12h": 43200,
    "1d": 86400,
}


class DatasetError(ValueError):
    """Dataset file violates the canonical-form contract."""


class KlineRow(BaseModel):
    """One dataset line; extra keys are a validation error (canonical form)."""

    model_config = ConfigDict(extra="forbid")

    symbol: SymbolStr
    interval: str
    open_time: int = Field(strict=True)
    open: DecimalStr
    high: DecimalStr
    low: DecimalStr
    close: DecimalStr
    volume: DecimalStr


def interval_ms(interval: str) -> int:
    """Candle duration in ms; unknown intervals are a validation error."""
    try:
        return INTERVAL_SECONDS[interval] * 1000
    except KeyError:
        raise DatasetError(f"unknown interval {interval!r}") from None


def close_time(open_time: int, d_ms: int) -> int:
    """close_time is DERIVED, never stored: ``open_time + d_ms``."""
    return open_time + d_ms


def grid_index(open_time: int, first_open_time: int, d_ms: int) -> int:
    """Grid index of a row: ``(open_time - first_open_time) // d_ms``."""
    return (open_time - first_open_time) // d_ms


def tick_count(rows: list[KlineRow]) -> int:
    """Total grid ticks N = last grid index + 1 (gapped indices own ticks)."""
    d_ms = interval_ms(rows[0].interval)
    return grid_index(rows[-1].open_time, rows[0].open_time, d_ms) + 1


def load_dataset(path: str | Path) -> tuple[list[KlineRow], str]:
    """Load and strictly validate a dataset file; returns (rows, dataset_sha256)."""
    raw = Path(path).read_bytes()
    sha256 = hashlib.sha256(raw).hexdigest()
    try:
        text = raw.decode("utf-8")
    except UnicodeDecodeError as exc:
        raise DatasetError(f"dataset is not UTF-8: {exc}") from exc
    rows: list[KlineRow] = []
    d_ms = 0
    for lineno, line in enumerate(text.splitlines(), start=1):
        if not line:
            raise DatasetError(f"dataset line {lineno}: blank line")
        try:
            payload = json.loads(line)
        except ValueError as exc:
            raise DatasetError(f"dataset line {lineno}: not valid JSON") from exc
        try:
            row = KlineRow.model_validate(payload)
        except ValidationError as exc:
            raise DatasetError(f"dataset line {lineno}: {exc}") from exc
        if not rows:
            d_ms = interval_ms(row.interval)
        else:
            first = rows[0]
            if row.symbol != first.symbol or row.interval != first.interval:
                raise DatasetError(
                    f"dataset line {lineno}: ({row.symbol}, {row.interval}) does not "
                    f"match dataset ({first.symbol}, {first.interval})"
                )
            if row.open_time <= rows[-1].open_time:
                raise DatasetError(
                    f"dataset line {lineno}: open_time {row.open_time} not strictly "
                    f"ascending (previous {rows[-1].open_time})"
                )
            if (row.open_time - first.open_time) % d_ms != 0:
                raise DatasetError(
                    f"dataset line {lineno}: open_time {row.open_time} off the "
                    f"{first.interval} grid anchored at {first.open_time}"
                )
        rows.append(row)
    if not rows:
        raise DatasetError("dataset file is empty")
    return rows, sha256


def closed_window(rows: list[KlineRow], t: int, window: int) -> list[KlineRow]:
    """Trailing min(window, available) rows with grid index <= t (the M0 mask path)."""
    if window < 1:
        raise ValueError(f"window must be >= 1, got {window}")
    first_open_time = rows[0].open_time
    d_ms = interval_ms(rows[0].interval)
    available = [
        row for row in rows if grid_index(row.open_time, first_open_time, d_ms) <= t
    ]
    return available[-window:]


def binance_klines(rows: list[KlineRow]) -> list[list[object]]:
    """Rows as Binance-style kline lists for the shared market_data formatter."""
    return [
        [
            row.open_time,
            decimal_to_str(row.open),
            decimal_to_str(row.high),
            decimal_to_str(row.low),
            decimal_to_str(row.close),
            decimal_to_str(row.volume),
        ]
        for row in rows
    ]

File: control-plane/internal/backtest/dataset.go
package backtest

import (
	"bufio"
	"bytes"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"os"

	"github.com/OmniMintX/AlphaMintX/control-plane/internal/contract"
)

// Kline is one dataset line. The struct field order IS the wire key order
// (encoding/json marshals in declaration order): the canonical dataset file
// must be byte-identical on both planes. open_time is ms since the Unix
// epoch; close_time is DERIVED (open_time + interval), never stored. Prices
// and volume are verbatim decimal strings (ADR-0003).
type Kline struct {
	Symbol   string `json:"symbol"`
	Interval string `json:"interval"`
	OpenTime int64  `json:"open_time"`
	Open     string `json:"open"`
	High     string `json:"high"`
	Low      string `json:"low"`
	Close    string `json:"close"`
	Volume   string `json:"volume"`
}

// Dataset is a validated, ordered candle series for one (symbol, interval).
// Klines may be GAPPED (missing grid slots); tick numbers index the full
// grid from the first candle, so a gap is a tick with no candle.
type Dataset struct {
	Symbol   string
	Interval string
	SHA256   string // hex sha256 of the exact file bytes
	Klines   []Kline
}

// IntervalMS is the candle duration in ms.
func (d *Dataset) IntervalMS() int64 {
	return intervalSeconds[d.Interval] * 1000
}

// FirstOpenTime is the grid origin (ms).
func (d *Dataset) FirstOpenTime() int64 { return d.Klines[0].OpenTime }

// Ticks is the number of decision ticks: one per grid slot from the first
// through the last candle, INCLUDING gapped slots.
func (d *Dataset) Ticks() int {
	last := d.Klines[len(d.Klines)-1].OpenTime
	return int((last-d.FirstOpenTime())/d.IntervalMS()) + 1
}

// TickOf is the grid index of a candle open_time.
func (d *Dataset) TickOf(openTime int64) int {
	return int((openTime - d.FirstOpenTime()) / d.IntervalMS())
}

// ReadDataset reads and validates a canonical dataset file, recording the
// sha256 of its exact bytes.
func ReadDataset(path string) (*Dataset, error) {
	b, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	sum := sha256.Sum256(b)
	ds := &Dataset{SHA256: hex.EncodeToString(sum[:])}
	scanner := bufio.NewScanner(bytes.NewReader(b))
	scanner.Buffer(make([]byte, 0, 64*1024), 4*1024*1024)
	lineNo := 0
	for scanner.Scan() {
		lineNo++
		line := scanner.Bytes()
		if len(line) == 0 {
			return nil, fmt.Errorf("dataset %s line %d: empty line", path, lineNo)
		}
		var k Kline
		dec := json.NewDecoder(bytes.NewReader(line))
		dec.DisallowUnknownFields()
		if err := dec.Decode(&k); err != nil {
			return nil, fmt.Errorf("dataset %s line %d: %w", path, lineNo, err)
		}
		if err := ds.appendValidated(k); err != nil {
			return nil, fmt.Errorf("dataset %s line %d: %w", path, lineNo, err)
		}
	}
	if err := scanner.Err(); err != nil {
		return nil, err
	}
	if len(ds.Klines) == 0 {
		return nil, fmt.Errorf("dataset %s: no klines", path)
	}
	return ds, nil
}

// appendValidated enforces the dataset invariants line by line: one
// (symbol, interval) per file, parseable decimals, strictly ascending
// grid-aligned open times.
func (ds *Dataset) appendValidated(k Kline) error {
	if len(ds.Klines) == 0 {
		if !symbolPattern.MatchString(k.Symbol) {
			return fmt.Errorf("symbol %q is not canonical BASE/QUOTE", k.Symbol)
		}
		if _, err := IntervalSeconds(k.Interval); err != nil {
			return err
		}
		ds.Symbol, ds.Interval = k.Symbol, k.Interval
	} else {
		if k.Symbol != ds.Symbol || k.Interval != ds.Interval {
			return fmt.Errorf("(%s, %s) does not match dataset (%s, %s)", k.Symbol, k.Interval, ds.Symbol, ds.Interval)
		}
		prev := ds.Klines[len(ds.Klines)-1].OpenTime
		if k.OpenTime <= prev {
			return fmt.Errorf("open_time %d not strictly ascending (previous %d)", k.OpenTime, prev)
		}
		if (k.OpenTime-ds.FirstOpenTime())%ds.IntervalMS() != 0 {
			return fmt.Errorf("open_time %d off the %s grid anchored at %d", k.OpenTime, ds.Interval, ds.FirstOpenTime())
		}
	}
	for _, f := range []struct{ name, v string }{
		{"open", k.Open}, {"high", k.High}, {"low", k.Low}, {"close", k.Close}, {"volume", k.Volume},
	} {
		// The strict ADR-0003 form (no signs, exponents, or leading zeros) —
		// the same rule the Python plane's DecimalStr enforces, so a file can
		// never pass one plane and fail the other.
		if _, err := contract.ParseDecimal(f.v); err != nil {
			return fmt.Errorf("%s %q: %w", f.name, f.v, err)
		}
	}
	ds.Klines = append(ds.Klines, k)
	return nil
}

// WriteDataset writes klines as the canonical dataset form: one compact-JSON
// line per candle in the pinned key order, LF-terminated. It returns the
// sha256 hex of the bytes written.
func WriteDataset(w io.Writer, klines []Kline) (string, error) {
	var buf bytes.Buffer
	ds := &Dataset{}
	for i, k := range klines {
		if err := ds.appendValidated(k); err != nil {
			return "", fmt.Errorf("kline %d: %w", i, err)
		}
		b, err := json.Marshal(k)
		if err != nil {
			return "", err
		}
		buf.Write(b)
		buf.WriteByte('\n')
	}
	if len(klines) == 0 {
		return "", fmt.Errorf("no klines to write")
	}
	sum := sha256.Sum256(buf.Bytes())
	if _, err := w.Write(buf.Bytes()); err != nil {
		return "", err
	}
	return hex.EncodeToString(sum[:]), nil
}
